src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1511 @@
+            Symbol sym = TreeInfo.symbolFor(tree);
+            if (sym != null &&
+                sym.kind == MTH &&
+                (sym.flags() & Flags.STATIC) == 0) {
+                type = sym.owner.type;
+            }
+
+            // If the expression type is incompatible with 'this', discard it
+            if (type != null && !targetClass.sym.isSubClass(type.tsym, types))
+                refs.remove(ExprRef.direct(depth));
+        }
+    }
+
+//
+// Visitor methods - Class Declarations
+//
+
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        return;     // we're busy analyzing another class - skip
+    }
+
+//
+// Visitor methods - Variable Declarations
+//
+
+    @Override
+    public void visitVarDef(JCVariableDecl tree) {
+
+        // Skip if ignoring warnings for this field
+        if (suppressed.contains(tree.sym))
+            return;
+
+        // Scan initializer, if any
+        scan(tree.init);
+        if (isParamOrVar(tree.sym))
+            refs.replaceExprs(depth, direct -> new VarRef(tree.sym, direct));
+        else
+            refs.discardExprs(depth);           // we don't track fields yet
+    }
+
+//
+// Visitor methods - Methods
+//
+
+    @Override
+    public void visitMethodDef(JCMethodDecl tree) {
+        Assert.check(false);        // we should never get here
+    }
+
+    @Override
+    public void visitApply(JCMethodInvocation invoke) {
+
+        // Get method symbol
+        MethodSymbol sym = (MethodSymbol)TreeInfo.symbolFor(invoke.meth);
+
+        // Recurse on method expression
+        scan(invoke.meth);
+        boolean direct = refs.remove(ExprRef.direct(depth));
+        boolean indirect = refs.remove(ExprRef.indirect(depth));
+
+        // Determine if method receiver represents a possible reference
+        RefSet<ThisRef> receiverRefs = RefSet.newEmpty();
+        if (sym != null && !sym.isStatic()) {
+            if (direct)
+                receiverRefs.add(ThisRef.direct());
+            if (indirect)
+                receiverRefs.add(ThisRef.indirect());
+        }
+
+        // If "super()": ignore - we don't try to track into superclasses.
+        // However, we do need to "invoke" non-static initializers/blocks.
+        Name name = TreeInfo.name(invoke.meth);
+        if (name == names._super) {
+            scanInitializers();
+            return;
+        }
+
+        // "Invoke" the method
+        invoke(invoke, sym, invoke.args, receiverRefs);
+    }
+
+    private void invoke(JCTree site, MethodSymbol sym, List<JCExpression> args, RefSet<?> receiverRefs) {
+
+        // Skip if ignoring warnings for a constructor invoked via 'this()'
+        if (suppressed.contains(sym))
+            return;
+
+        // Ignore final methods in java.lang.Object (getClass(), notify(), etc.)
+        if (sym != null &&
+            sym.owner.kind == TYP &&
+            ((ClassSymbol)sym.owner).fullname == names.java_lang_Object &&
```

Comment:
nit: in general we use another idiom for this in the compiler:
```
sym.owner.type.tsym == syms.objectType.tsym
```
where `syms` is an instance of: `com.sun.tools.javac.code.Symtab`

Reply:
Thanks - will fix.